presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBufferFactory.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.buffer;
+
+import com.facebook.presto.execution.StateMachine;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.spiller.LocalTempStorage;
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.facebook.presto.storage.TempStorageManager;
+import com.facebook.presto.testing.TestingTempStorageManager;
+import com.facebook.presto.util.FinalizerService;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.io.Closer;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.inject.Inject;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+
+public class SpoolingOutputBufferFactory
+{
+    private ListeningExecutorService executor;
+
+    private final FeaturesConfig featuresConfig;
+    private final TempStorageManager tempStorageManager;
+    private final FinalizerService finalizerService;
+
+    private final Closer closer = Closer.create();
+
+    private static ExecutorService coreExecutor = newCachedThreadPool(daemonThreadsNamed("spooling-outputbuffer-%s"));
+
+    @Inject
+    public SpoolingOutputBufferFactory(FeaturesConfig featuresConfig, TempStorageManager tempStorageManager, FinalizerService finalizerService)
+    {
+        this.featuresConfig = requireNonNull(featuresConfig, "featuresConfig is null");
+        this.tempStorageManager = requireNonNull(tempStorageManager, "tempStorageManger is null");
+        this.finalizerService = requireNonNull(finalizerService, "finalizerService is null");
+    }
+
+    @VisibleForTesting
+    public SpoolingOutputBufferFactory(FeaturesConfig featuresConfig)
+    {
+        this.featuresConfig = requireNonNull(featuresConfig);
+        tempStorageManager = new TestingTempStorageManager();
+        finalizerService = new FinalizerService();
+
+        initialize();
```

Comment:
do you have to call initialize() here when it's already annotated with @PostConstruct?

Reply:
Yes. I believe `initialize` is only called after the dependency injection with @PostConstruct. This constructor is used for testing and doesn't have any injection, so I need to call initialize for the executor to work.

Reply:
Can we use the above constructor and just manually pass in `TestingTempStorageManager` and create the new `FinalizerService`?

Reply:
No, even though the constructor above is marked with @Inject, no injection is actually being done when I create the `SpoolingOutputBufferFactory`, so initialize still won't get called.